clientapi/routing/sso.go

```diff
@@ -0,0 +1,203 @@
+			JSON: jsonerror.NotFound("Method disabled"),
+		}
+	}
+
+	// A redirect URL is required for this endpoint
+	redirectURLStr := req.FormValue("redirectUrl")
+	if redirectURLStr == "" {
+		return util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.MissingArgument("redirectUrl parameter missing"),
+		}
+	}
+	// Check if the redirect url is a valid URL
+	redirectURL, err := url.Parse(redirectURLStr)
+	if err != nil {
+		return util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.InvalidArgumentValue("Invalid redirectURL: " + err.Error()),
+		}
+	}
+
+	// If the request has a ticket param, validate the ticket instead of redirecting to SSO server
+	if ticket := req.FormValue("ticket"); ticket != "" {
+		return ssoTicket(req, redirectURL, accountDB, cfg)
+	}
+
+	// Adding the params to the sso url
+	ssoURL := cfg.CAS.URL
+	ssoQueries := make(url.Values)
+	// the service url that we send to CAS is homeserver.com/_matrix/client/r0/login/sso/redirect?redirectUrl=xyz
+	ssoQueries.Set("service", req.RequestURI)
+	ssoURL.RawQuery = ssoQueries.Encode()
+
+	return util.RedirectResponse(ssoURL.String())
+}
+
+// ssoTicket handles the m.login.sso login attempt after the user had completed auth at the SSO server
+// - gets the ticket from the SSO server (this is different from the matrix login/access token)
+// - calls validateTicket to validate the ticket
+// - calls completeSSOAuth
+func ssoTicket(
+	req *http.Request,
+	redirectURL *url.URL,
+	accountDB accounts.Database,
+	cfg *config.ClientAPI,
+) util.JSONResponse {
+	// form the ticket validation URL from the config
+	validateURL := cfg.CAS.ValidateURL
+	ticket := req.FormValue("ticket")
+
+	// append required params to the CAS validate endpoint
+	validateQueries := make(url.Values)
+	validateQueries.Set("ticket", ticket)
+	validateURL.RawQuery = validateQueries.Encode()
+
+	// validate the ticket
+	casUsername, err := validateTicket(validateURL.String())
+	if err != nil {
+		// TODO: should I be logging these? What else should I log?
+		util.GetLogger(req.Context()).WithError(err).Error("CAS SSO ticket validation failed")
+		return util.JSONResponse{
+			Code: http.StatusUnauthorized,
+			JSON: jsonerror.Unknown("Could not validate SSO token: " + err.Error()),
+		}
+	}
+	if casUsername == "" {
+		util.GetLogger(req.Context()).WithError(err).Error("CAS SSO returned no user")
+		return util.JSONResponse{
+			Code: http.StatusUnauthorized,
+			JSON: jsonerror.Unknown("CAS SSO returned no user"),
+		}
+	}
+
+	// ticket validated. Login the user
+	return completeSSOAuth(req, casUsername, redirectURL, accountDB)
+}
+
+// validateTicket sends the ticket to the sso server to get it validated
+// the CAS server responds with an xml which contains the username
+// validateTicket returns the SSO User
+func validateTicket(
+	ssoURL string,
+) (string, error) {
+	// make the call to the sso server to validate
+	response, err := http.Get(ssoURL)
```

Comment:
Please use an HTTP client here, as the default `http.Get` has [no timeout](https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779).